user/service.py

```diff
@@ -1,11 +1,42 @@
+import hashlib, uuid
+
+import user.dao as user_dao
+
 def get(user_id: int = None) -> list | None:
-    return None
+    """Get a user by ID, or all users if no ID is given."""
+
+    if user_id is not None:
+        return user_dao.get_by_id(user_id)
+    else:
+        return user_dao.get_all()
+
+def hash_password(salted_password: str) -> str:
+    """Return the hash for a password that has been salted."""
+    return hashlib.sha256(salted_password.encode()).hexdigest()
+
+def generate_salt() -> str:
+    """Generate a random salt."""
+    return str(uuid.uuid4())
 
 def create(username: str = None, password: str = None) -> int:
-    return -1
+    """Create a new user."""
+    if user_dao.get_by_username(username) is None:
+        return -1
+
+    salt = generate_salt()
+    return user_dao.create(username, hash_password(password + salt), salt)
 
-def update(user_id: int, email: str = None, password: str = None) -> bool:
-    return False
+def update(user_id: int, username: str = None, password: str = None) -> bool:
+    """Update a user given its ID."""
+    if user_dao.get_by_id(user_id) is None:
+        return False
+
+    salt = user_dao.get_salt(user_id)
```

Comment:
this should be totally fine but at some point I think DB/Backend teams need to have some sort of discussion about whether the DAO module should cache results or make separate requests to the DB for each of these calls, and in the latter case we may way to switch this to something like

```python
user = user_dao.get_by_id(user_id)
if user is None:
    return False

salt = user.salt # depends on the interface, but this should be possible with the way we're implementing the DAOs right now
```

Reply:
just saw the changes and that looks good to me! as it stands right now, the DAOs do not have getter methods (unless mongoengine secretly adds them? I need to test that) so we'll probably have to add them.